### PR TITLE
KAFKA-7223: internally provide full consumer record during restore

### PR DESCRIPTION
**Pull request by vvcephei**
The Suppression buffer stores the full record context, not just the key and value, 
so its changelog/restore loop will also need to preserve this information.

This change is a precondition to that, creating an option to register a 
state restore callback to receive the full consumer record. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
@guozhangwang @bbejeck @mjsax ,

This is some groundwork in support of the upcoming Suppression Part 4 pr.

As I said in the description, unlike other state stores, the suppression buffer needs to symmetrically restore *all* the record context that it stores in the Record Collector.

We could consider offering this as a public API in some form for state stores in the future, but for now, I have taken care to keep the change internal.

This change should also support, e.g., KIP-258.

Please let me know what you think!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/BatchingStateRestoreRecordCallback.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/BatchingStateRestoreRecordCallback.java]: This is the crux of the whole changeset: a callback that gives access to the whole context.

It doesn't really have to be the `ConsumerRecord` itself; I could also define a new class to contain the data+context, but ConsumerRecord happens to contain exactly what we need already. So I figured I'd keep it simple for internal use.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/BatchingStateRestoreRecordCallback.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/BatchingStateRestoreRecordCallback.java]: I have no clue why Git thinks this was a rename; It's a completely different class.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:280]: The way it works is that now, we just go ahead and send the full consumer records, and the adapter will take care of converting it to `KeyValue`s if that's what the callback needs.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java]: I consolidated the prior `WrappedBatchingStateRestoreCallback`, as well as an independent `getBatchingRestoreCallback` into this method.

This way, the logic is more centralized, and we also avoid creating an extra wrapper class if the `restoreCallback` is already of the desired type.

File: streams/src/test/java/org/apache/kafka/test/InternalMockProcessorContext.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/test/InternalMockProcessorContext.java:308]: We actually never should have been adapting it to a `BatchingStateRestoreCallback` before casting it to a `StateRestoreListener`.

The only reason it worked is that all our implementaitons happen to be `BatchingStateRestoreCallback` already, and the adaptation was therefore a no-op.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/BatchingStateRestoreRecordCallback.java
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/BatchingStateRestoreRecordCallback.java]: I think this may have been removed by accident.  The PR doesn't contain `WrappedBatchingStateRestoreCallback` but the class exists in trunk.

EDIT: NM I see what is going on further down in the PR

**Review by bbejeck — COMMENTED** [review]
Thanks for the patch @vvcephei.  Overall this looks good.  I have a few meta-comments.

1.  I could be missing something, but I don't see from this PR where we need the `ConsumerRecord` vs the previous version using `KeyValue<byte[], byte[]>`.  Is the need for the info in the `ConsumerRecord` coming in a follow-on PR?
2. To me, the naming of `BatchingStateRestoreRecordCallback` and `BatchingStateRestoreCallback` is a little similar, I kept getting the two confused.  I don't have any good suggestions ATM maybe call the new interface `RecordBatchingStateRestoreCallback`? I don't know if that would be an improvement or not.
3. IMHO we could add two additional tests to `StateRestoreCallbackAdapterTest` asserting the two possible callbacks did get the records as expected.  I realize we have this covered from other tests, but my personal preference is to have some separate independent tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/BatchingStateRestoreRecordCallback.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/BatchingStateRestoreRecordCallback.java]: yeah, I removed this class and added an unrelated interface. I guess the file structure was similar enough that Git decided it was a rename?

**Comment by vvcephei** [comment]
Thanks for the review @bbejeck.
1. Yes, it'll be needed in Suppression Part 4. 
FWIW, the code path isn't dead in this PR, since the adapter converts all the other callbacks into the new one. So, although it's not used explicitly, it's being exercised by all the other tests.
2. I like your name better. I'll update it.
3. Roger that. I'll add them.

**Comment by vvcephei** [comment]
@bbejeck I've addressed your comments.

**Comment by vvcephei** [comment]
Looks like the jenkins runner ran out of disk...

Retest this, please.

**Comment by vvcephei** [comment]
java 8 passed that time, but java 10 was still out of disk.

Retest this, please.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java:30]: nit: can we omit this constructor since JDK will create the default one just like this right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:273]: We can just use `ConsumerRecords#records(TopicPartition partition)` since it should only contain one topic-partition's records.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java:47]: If `restoreCallback` is null we will throw NPE here right?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapterTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapterTest.java:61]: nit: create a single list from `asList` for restore and validation?

**Review by guozhangwang — COMMENTED** [review]
LGTM overall, just one major comment around NPE guarding, others are minor.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java:30]: This privatizes the constructor to ensure the class can't be instantiated (thus only used as a container for static methods)

**Review by bbejeck — APPROVED** [review]
Thanks @vvcephei LGTM modulo @guozhangwang comments.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java:47]: We would throw the NPE later on when `restoreRecords` is actually called, which is what currently happens with the `WrappedBatchingStateRestoreCallback`.

Would you feel better to surface the NPE immediately with a check at the beginning of `adapt`?

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java:47]: I think this would be better, unless there's some semantics around registering a null callback.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java:30]: Ack

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java:47]: Yeah I felt it is better to guard against null beforehand. This actually come from the unit test where we do `adapt(null).restore(null)` where the `restore` function is supposed to throw not-supported exception, while I was a bit frowning upon it.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:273]: I'm sorry; I didn't follow this. 

Are you suggesting to do `final List<ConsumerRecord<byte[], byte[]>> restoreRecords = records.records(topicPartition)`? 

This would return the right type, but `restoreRecords`  is a filtered list of only records with non-null keys. It doesn't look like `ConsumerRecords#records` applies this filter.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StateRestoreCallbackAdapter.java:47]: ok, it's done now.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:273]: I need to be specific with my comments :)

I'm suggesting sth like `records.records(topicPartition).stream().filter().collect()`, the main point is that `ConsumerRecords#iterate` which is used as `final ConsumerRecord<byte[], byte[]> record : records` is a pretty complex operator and is unnecessary here since we know it will only contain one topic-partition.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:273]: aaah. Now I understand.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:194]: I'm not sure what this test was previously checking for. It looks like it wanted to try updating without previously initializing the task.

This was previously throwing an NPE wrapped in a ProcessorStateException when the adapted callback actually tried to dereference the (null) wrapped callback.

Now, the NPE is surfaced a little earlier, not guarded by the ProcessorStateException try/catch.

I think this is still ok, so I've just refined the test to be more specific.

**Comment by vvcephei** [comment]
Ok, @guozhangwang, I think I've addressed all your comments.
Thanks for the review, by the way!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:273]: I've done this, but I stuck with the foreach loop (and also went back to it in the adapter), as this code path is hot, and I'm not sure of the overhead of `stream()`.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:273]: In that case could we just use the foreach loop after `ConsumerRecords#records` to get the filter list from the returned list? I just felt leveraging on `ConsumerRecords#iterator` is unnecessarily costly.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java
Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:194]: This test case is added in the first commit of this test file, and after reading it I agree with you that the original name `testUpdateNonPersistentStore` seems make less sense, and rather `testUpdateNonInitializedStore` is better.

**Comment by guozhangwang** [comment]
@vvcephei one follow-up comment https://github.com/apache/kafka/pull/5710#discussion_r221415280 otherwise lgtm.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java:273]: I think this is what I've done in the most recent version of the code: https://github.com/apache/kafka/pull/5710/files#diff-e870eab2dedeee93442ba87dc12609a4R277
```
for (final ConsumerRecord<byte[], byte[]> record : records.records(topicPartition)) {
  if (record.key() != null) {
    restoreRecords.add(record);
  }
}
```
Is that what you're thinking?

**Review by mjsax — APPROVED** [review]
LGTM.

**Comment by vvcephei** [comment]
java 10 passed; java 8 test failure unrelated.

Retest this, please.

**Comment by mjsax** [comment]
Retest this please.

If Java10 keeps running out of disk space, should be fine to merge with Java8 passing IMHO. Leave it to @guozhangwang to merge.

**Pull request merged**
